flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,562 @@
+
+            modelDataState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>("modelData", KMeansModelData.class));
+        }
+
+        @Override
+        public void processElement1(StreamRecord<DenseVector[]> pointsRecord) throws Exception {
+            miniBatchState.add(pointsRecord.getValue());
+            processElement();
+        }
+
+        @Override
+        public void processElement2(
+                StreamRecord<Tuple2<KMeansModelData, DenseVector>> modelDataAndWeightsRecord)
+                throws Exception {
+            modelDataState.add(modelDataAndWeightsRecord.getValue().f0);
+            processElement();
+        }
+
+        private void processElement() throws Exception {
+            if (!modelDataState.get().iterator().hasNext()
+                    || !miniBatchState.get().iterator().hasNext()) {
+                return;
+            }
+
+            DenseVector[] centroids = getUniqueElement(modelDataState, "modelData").centroids;
+            modelDataState.clear();
+
+            List<DenseVector[]> pointsList = IteratorUtils.toList(miniBatchState.get().iterator());
+            DenseVector[] points = pointsList.get(0);
+            pointsList.remove(0);
+            miniBatchState.clear();
+            miniBatchState.addAll(pointsList);
+
+            // Computes new centroids.
+            DenseVector[] sums = new DenseVector[k];
+            int dim = centroids[0].size();
+            DenseVector counts = new DenseVector(k);
+
+            for (int i = 0; i < k; i++) {
+                sums[i] = new DenseVector(dim);
+                counts.values[i] = 0;
+            }
+            for (DenseVector point : points) {
+                int closestCentroidId =
+                        KMeans.findClosestCentroidId(centroids, point, distanceMeasure);
+                counts.values[closestCentroidId]++;
+                for (int j = 0; j < dim; j++) {
+                    sums[closestCentroidId].values[j] += point.values[j];
+                }
+            }
+
+            for (int i = 0; i < k; i++) {
+                if (counts.values[i] < 1e-5) {
+                    continue;
+                }
+                BLAS.scal(1.0 / counts.values[i], sums[i]);
+            }
+
+            output.collect(new StreamRecord<>(Tuple2.of(new KMeansModelData(sums), counts)));
+        }
+    }
+
+    private static class FeaturesExtractor implements MapFunction<Row, DenseVector> {
+        private final String featuresCol;
+
+        private FeaturesExtractor(String featuresCol) {
+            this.featuresCol = featuresCol;
+        }
+
+        @Override
+        public DenseVector map(Row row) throws Exception {
+            return (DenseVector) row.getField(featuresCol);
+        }
+    }
+
+    private static class MiniBatchDistributor
```

Comment:
Got it. And since the term "mini batch" would cause ambiguity according to other discussions, I'll rename the operators to `GlobalBatchCreator` and `LocalBatchDistributor`.